openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -188,30 +201,90 @@ def assert_response_with_delete_cookie(
             self.assert_response(set_request_user=set_request_user, set_session_cookie=set_session_cookie)
             assert mock_delete_cookie.called == expect_delete_called
 
-    def test_success(self):
+    @patch("django.core.cache.cache.set")
+    @patch("django.core.cache.cache.touch")
+    def test_success(self, mock_cache_set, mock_cache_touch):
         with self.assert_not_logged():
             self.assert_response(set_request_user=True, set_session_cookie=True)
+        assert not mock_cache_set.called
+        assert not mock_cache_touch.called
 
-    def test_confirm_user_at_step_2(self):
+    @patch("django.core.cache.cache.set")
+    @patch("django.core.cache.cache.touch")
+    def test_confirm_user_at_step_2(self, mock_cache_set, mock_cache_touch):
         self.request.safe_cookie_verified_user_id = self.user.id
         with self.assert_not_logged():
             self.assert_response(set_request_user=True, set_session_cookie=True)
-
-    def test_different_user_at_step_2_error(self):
-        self.request.safe_cookie_verified_user_id = "different_user"
-
-        with self.assert_logged_for_request_user_mismatch("different_user", self.user.id, 'warning', self.request.path):
-            self.assert_response(set_request_user=True, set_session_cookie=True)
-
-        with self.assert_logged_for_session_user_mismatch("different_user", self.user.id, self.request.path):
-            self.assert_response(set_request_user=True, set_session_cookie=True)
-
-    def test_anonymous_user(self):
+        assert not mock_cache_set.called
+        assert not mock_cache_touch.called
+
+    @patch("django.core.cache.cache.set")
+    @ddt.unpack
+    @ddt.data(
+        {'change_response_user': True, 'change_session_user': False, 'mismatch_location': 'response',
+         'change_session': True},
+        {'change_response_user': False, 'change_session_user': True, 'mismatch_location': 'session',
+         'change_session': True},
+        {'change_response_user': True, 'change_session_user': True, 'mismatch_location': 'both',
+         'change_session': True},
+        {'change_response_user': True, 'change_session_user': False, 'mismatch_location': 'response',
+         'change_session': False},
+        {'change_response_user': False, 'change_session_user': True, 'mismatch_location': 'session',
+         'change_session': False},
+        {'change_response_user': True, 'change_session_user': True, 'mismatch_location': 'both',
+         'change_session': False},
+    )
+    def test_different_user_at_step_2_error(self, mock_cache_set, change_response_user,
+                                            change_session_user, mismatch_location, change_session):
+        new_user = UserFactory()
+        self.request.safe_cookie_verified_user_id = self.user.id
```

Comment:
I wonder if there is a way to mock slightly less in this test? Could we call the actual middleware on the request and have it do this instead, and then muck with stuff, and then call the middleware on the response?